code/datums/components/egg_layer.dm

```diff
@@ -0,0 +1,84 @@
+/**
+ * # egg layer component!
+ *
+ * Component that manages how many eggs to lay, what can be fed to the mob to make them lay more, and what is actually laid.
+ * Since the only real interaction with the component is an attackby, the nice part is that we're able to make this an atom level proc.
+ * egg_layer will loudly fail if you do not provide it the arguments, as to encourage explicicy(?)
+ */
+/datum/component/egg_layer
+	/// item laid by the mob
+	var/egg_type
+	/// items that can be fed to the mob to make it lay more eggs
+	var/list/food_types
+	/// messages sent when fed
+	var/list/feedMessages
+	/// messages sent when laying an egg
+	var/list/layMessages
+	/// kind of a weird boolean, but basically egg layers usually need to make their eggs process to let them grow. and some don't want that to happen too!
+	var/process_eggs
+	/// how many eggs left to lay
+	var/eggs_left = 0
+	/// used to check for possibly too much egg laying going on (aka past a certain amount of chickens, stop making more chickens from eggs)
+	var/datum/callback/egg_animal_callback
+
+/datum/component/egg_layer/Initialize(egg_type, food_types, feedMessages, layMessages, eggs_left, process_eggs, egg_animal_callback)
+	if(!isatom(parent)) //yes, you could make a tameable toolbox.
+		return COMPONENT_INCOMPATIBLE
+
+	src.egg_type = egg_type
+	src.food_types = food_types
+	src.feedMessages = feedMessages
+	src.layMessages = layMessages
+	src.eggs_left = eggs_left
+	src.process_eggs = process_eggs
+	src.egg_animal_callback = egg_animal_callback
+
+	RegisterSignal(parent, COMSIG_PARENT_ATTACKBY, .proc/feed_food)
```

Comment:
Unregister this signal in UnregisterFromParent (and register it in RegisterParent as well, or whatever the names are instead of Initialize)

Reply:
true that